lib/auth/keystore/gcp_kms.go

```diff
@@ -228,65 +228,96 @@ func (g *gcpKMSKeyStore) canSignWithKey(ctx context.Context, raw []byte, keyType
+// different Teleport cluster and the keys it's actively using will never appear
+// in the activeKeys argument
 func (g *gcpKMSKeyStore) DeleteUnusedKeys(ctx context.Context, activeKeys [][]byte) error {
 	// Make a map of currently active key versions, this is used for lookups to
-	// check which keys in KMS are unused, and holds a count of how many times
-	// they are found in KMS. If any active keys are not found in KMS, we are in
-	// a bad state, so key deletion will be aborted.
+	// check which keys in KMS are unused.
 	activeKmsKeyVersions := make(map[string]int)
 	for _, activeKey := range activeKeys {
-		keyID, err := parseGCPKMSKeyID(activeKey)
+		keyIsRelevant, err := g.canSignWithKey(ctx, activeKey, keyType(activeKey))
 		if err != nil {
-			// could be a different type of key
+			return trace.Wrap(err)
+		}
+		if !keyIsRelevant {
+			// Ignore active keys that are not GCP KMS keys or are in a
+			// different keyring than the one this Auth is configured to use.
 			continue
 		}
+		keyID, err := parseGCPKMSKeyID(activeKey)
+		if err != nil {
+			// Realistically we should not hit this since canSignWithKey already
+			// calls parseGCPKMSKeyID.
+			return trace.Wrap(err)
+		}
 		activeKmsKeyVersions[keyID.keyVersionName] = 0
 	}
 
-	var unusedKeyIDs []gcpKMSKeyID
+	var keysToDelete []gcpKMSKeyID
 
 	listKeyRequest := &kmspb.ListCryptoKeysRequest{
 		Parent: g.keyRing,
-		Filter: fmt.Sprintf("labels.%s=%s", hostLabel, g.hostUUID),
+		// Only bother listing keys created by Teleport which should have the
+		// hostLabel set
+		Filter: fmt.Sprintf("labels.%s:*", hostLabel),
 	}
 	iter := g.kmsClient.ListCryptoKeys(ctx, listKeyRequest)
 	key, err := iter.Next()
 	for err == nil {
 		keyVersionName := key.Name + keyVersionSuffix
-		if _, active := activeKmsKeyVersions[keyVersionName]; active {
-			activeKmsKeyVersions[keyVersionName]++
-		} else {
-			unusedKeyIDs = append(unusedKeyIDs, gcpKMSKeyID{
+		_, keyIsActive := activeKmsKeyVersions[keyVersionName]
+		if keyIsActive {
+			// Record that this current active key was actually found.
+			activeKmsKeyVersions[keyVersionName] += 1
+		} else if key.Labels[hostLabel] == g.hostUUID {
+			// This key is not active (it is not currently stored in any
+			// Teleport CA) and it was created by this Auth server, so it should
+			// be safe to delete.
+			keysToDelete = append(keysToDelete, gcpKMSKeyID{
 				keyVersionName: keyVersionName,
 			})
 		}
 		key, err = iter.Next()
 	}
 	if err != nil && !errors.Is(err, iterator.Done) {
-		return trace.Wrap(err)
+		return trace.Wrap(err, "unexpected error while iterating GCP KMS keys")
 	}
 
+	// If any member of activeKeys which is part of the same GCP KMS keyring
+	// queried here was not found in the ListCryptoKeys response, something has
+	// gone wrong and there's a chance we have a bug or GCP has made a breaking
+	// API change. In this case we should abort to avoid the chance of deleting
+	// any currently active keys.
```

Comment:
should we continue deleting unused keys? 

or is it possible to filter activeKeys by current hostUUID?

Reply:
Hmm I don't quite see what you're getting at here, can you elaborate? One piece of info I can offer is that we have no record of the hostUUID which created the key on the Teleport side, it's only stored as a label on the key in GCP

Reply:
Was thinking weird situations where an active key is not found in keyring, and it's not owned by this Auth. But reading the comment again `found == 0` only happens when something horribly went wrong? just ignore my previous comment =p